### ISSUE TITLE
Lints disappear a few seconds after having pressed Ctrl+S

### ISSUE THREAD

**Issue by kangalio** [rust-lang/rust-analyzer#8257]
Since recently, a weird problem has cropped up: after saving the file with Ctrl+S, rust-analyzer displays lints and errors in the code as usual. But after a few seconds, the displayed errors and lints just disappear entirely. When saving the file again, they reappear, but then disappear again after a few seconds.

When going through that cycle enough times, most of the time the lints eventually stay on the screen and don't disappear anymore.

I'm using rust-analyzer version v0.2.521 (as reported from inside VSCode extension menu; I can't tell the rust-analyzer reported version because its popup is blank). Server statistics: https://pastebin.com/3FBsqYaH.

I'm not sure what additional metadata could be gathered that is of interest here. I'll gladly provide more information about my setup and the environment.

**Comment by Veykril** [comment]
That's a really odd issue 😕 One good thing to post would your "Rust Analyzer Client" logs. Just to make sure is this issue happening with all extensions disabled except for rust analyzer?

**Comment by kangalioo** [comment]
Before seeing your message, I tried uninstall rust-analyzer, restart VSCode a few times and coding a bit if Rust without rust-analyzer in hopes to get VSCode to "recover". After reinstalling, the weird issue with the disappearing lints seems to have solved itself...

Sorry for the anticlimactic conclusion :sweat_smile:

**Comment by Veykril** [comment]
Well that's good then 😄

**Comment by kangalioo** [comment]
Nevermind, it reappeared after a few hours. I rebuilt the vscode config (deleted `~/.config/Code - OSS/User/*.json` and only reimplemented the settings tweaks I need for development), disabled all other extensions, restarted vscode, and tried writing some code. But soon enough, the issue cropped up again.

Rust Analyzer Client: https://pastebin.com/raw/QpVAVWWv

**Comment by Veykril** [comment]
does this issue still happen for you?

**Comment by kangalioo** [comment]
I've kept Check On Save disabled so not sure. I just did a very quick and dirty check and the lints stayed. I'll keep Check On Save enabled for now and reopen this issue if needed

**Comment by JuliDi** [comment]
This issue seems to be still present. Everything behaves as expected but after a few seconds the dimming and some errors in the "Problems" tab disappear.
![Screenshot 2023-04-22 at 13 40 19](https://user-images.githubusercontent.com/20155974/233782712-2c9b91fd-6de6-4fcc-a994-a92a00c7d75c.png)

I would expect these two lines to stay dimmed/grayed since the std feature is not set. But after a few seconds I either get no dimming and no error or I only get some "blahtest" not defined error.
The rust-analyzer client logs show no errors or so.

**Comment by JuliDi** [comment]
Maybe this helps in some way: For the first few seconds, I have dimmed inactive code and some unexpanded macro warnings. Then somehing happens (build script executed? no idea...) and the errors go away. But so goes the dimming :/
![Screenshot 2023-04-22 at 14 04 32](https://user-images.githubusercontent.com/20155974/233783709-11662011-e313-43ad-a8a8-12bbf4dc3adf.png)

**Comment by heksesang** [comment]
I have just run into a problem that sounds just like this. Figured I'd just explain my issue here in case this is the same thing.

I am using Zed, and when I save the file it runs `cargo check`. While `cargo check` runs it displays the lints, then they disappear just as soon as `cargo check` completes. For some reason pure syntax errors stay after the `cargo check`, while errors like using a variable that does not exist disappear.

From the RPC messages it looks like it starts sending `$/progress` messages for flycheck, sends `textDocument/publishDiagnostics` messages with diagnostics, then after flycheck ends it sends a new `textDocument/publishDiagnostics` message with empty diagnostics. From my understanding of the LSP specification, that last message should wipe the diagnostics from the editor, so I am not sure why it is even being sent?

**Comment by ShoyuVanilla** [comment]
> I have just run into a problem that sounds just like this. Figured I'd just explain my issue here in case this is the same thing.
> 
> I am using Zed, and when I save the file it runs `cargo check`. While `cargo check` runs it displays the lints, then they disappear just as soon as `cargo check` completes. For some reason pure syntax errors stay after the `cargo check`, while errors like using a variable that does not exist disappear.
> 
> From the RPC messages it looks like it starts sending `$/progress` messages for flycheck, sends `textDocument/publishDiagnostics` messages with diagnostics, then after flycheck ends it sends a new `textDocument/publishDiagnostics` message with empty diagnostics. From my understanding of the LSP specification, that last message should wipe the diagnostics from the editor, so I am not sure why it is even being sent?

Which version are you on? I guess this might be https://github.com/rust-lang/rust-analyzer/issues/20427?

**Comment by heksesang** [comment]
I am on `rust-analyzer 0.3.2571-standalone (4e147e7879 2025-08-10)`. So it might be #20427 as you say; I'll try upgrade to a version where this is fixed.

**Comment by heksesang** [comment]
Installing the latest stable seemed to solve my issue. And this issue and #20427 aren't related?

**Comment by ShoyuVanilla** [comment]
Yes, #20427 happened quite recently for `0.3.2571-standalone` due to my mistake and this one is much older.
